Add check for symlinks pointing outside of repo

mkosiarc opened the pull request:
If the check is enabled and the cloned repository contains symlinks pointing outside of the repo, the build will fail.

[STONEBLD-1328](https://issues.redhat.com//browse/STONEBLD-1328)

MartinBasti commented on task/git-clone/0.1/git-clone.yaml:
```diff
@@ -188,6 +207,15 @@ spec:
+
+      if [ "${PARAM_ENABLE_SYMLINK_CHECK}" = "true" ] ; then
+        check_symlinks
+        EXIT_CODE="$?"
+        if [ "${EXIT_CODE}" != 0 ] ; then
+          exit "${EXIT_CODE}"
+        fi
```
Isn't this behavior alredy enforced by
```
set -eu
```

MartinBasti replied:
and because you are using pipe in your code, it should be probably extended to
```
set -euo pipefail
```

mkosiarc replied:
Updated code and changed sh to bash so I can use process substitution.

Michkov commented on task/git-clone/0.1/git-clone.yaml:
```diff
@@ -165,6 +172,18 @@ spec:
+          do
+            target=$(readlink -f "$symlink")
+            if ! [[ "$target" =~ ^$CHECKOUT_DIR ]]; then
+              echo "The cloned repository contains symlink pointing outside of the cloned repository: $symlink"
```
would be better to aggregate all symlinks pointing outside and print all of them not just first one.

mkosiarc replied:
changed

tisutisu commented on task/git-clone/0.1/git-clone.yaml:
```diff
@@ -165,6 +172,18 @@ spec:
       }
 
+      check_symlinks() {
+        find $CHECKOUT_DIR -type l -print | \
```
Parent image does not contain the find command .. so this command fails
`+ find /workspace/output/ -type l -print
/tekton/scripts/script-0-7m9xq: line 39: find: command not found`

mkosiarc replied:
now the check is in different step that uses different image that has the find command available.

mkosiarc commented:
tested this with the following repo contaiting two symlinks pointing outside of the directory and one valid symlink https://github.com/mkosiarc-rhtap/symlink-check-test

mkosiarc commented:
> It's hard to review with all the indentation changes

I am aware, but since I added another step I had to indent it another level

Michkov commented:
> > It's hard to review with all the indentation changes
> 
> I am aware, but since I added another step I had to indent it another level

I do not agree, you have not added another level, just added new item to list. If you want to update indentation then please do that in separate commit.

mkosiarc commented:
> > > It's hard to review with all the indentation changes
> > 
> > 
> > I am aware, but since I added another step I had to indent it another level
> 
> I do not agree, you have not added another level, just added new item to list. If you want to update indentation then please do that in separate commit.

You are right, I misused my local linter. I apologize for the confusion and now the commit is updated.

Michkov commented on task/git-clone/0.1/git-clone.yaml:
```diff
@@ -190,6 +195,42 @@ spec:
       printf "%s" "${RESULT_SHA}" > "$(results.commit.path)"
       printf "%s" "${PARAM_URL}" > "$(results.url.path)"
+  - name: symlink-check
+    image: registry.redhat.io/ubi8:8.8-1009
```
probably better to use ubi9.

mkosiarc replied:
changed

Michkov commented on task/git-clone/0.1/git-clone.yaml:
```diff
@@ -190,6 +195,42 @@ spec:
+      value: $(workspaces.output.path)
+    resources: {}
+    securityContext:
+      runAsUser: 0
```
nit: this is not needed since ubi is using root(uid 0) by default.

mkosiarc replied:
changed
